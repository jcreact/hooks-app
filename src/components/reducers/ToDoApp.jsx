import React, { useEffect, useReducer } from 'react';

//Componentes
import { ToDoAdd } from './ToDoAdd.jsx';
import { ToDoList } from './ToDoList.jsx';

// Reducer
import { todoReducer } from './todoReducer';

const initializerToDos = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];
};

export const ToDoApp = () => {
    const [todos, dispatch] = useReducer(todoReducer, [], initializerToDos);

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

    /**
     *
     * @param {{id: number, project:string, task:string, done:boolean}} todo - Nueva tarea.
     */
    const handleAdd = (todo) => {
        dispatch({
            type: 'add',
            payload: todo,
        });
    };

    /**
     * Elimina una tarea de la lista.
     * @param {number} id - Identificardor de la tarea.
     */
    const removeTask = (id) => {
        dispatch({
            type: 'remove',
            payload: id,
        });
    };

    /**
     * Cambia el estado de la tarea.
     * @param {number} id - Identificador de la tarea.
     */
    const handleToggle = (id) => {
        dispatch({
            type: 'toggle',
            payload: id,
        });
    };

    return (
        <div className="container-fluid">
            <div className="mt-3 d-flex justify-content-start align-items-start">
                <h1>To-Do App</h1>
                <span className="badge bg-danger rounded-pill">
                    {todos.filter((t) => !t.done).length}
                </span>
            </div>
            <hr />

            <ToDoAdd handleAdd={handleAdd} />

            <ToDoList
                todos={todos}
                removeTask={removeTask}
                handleToggle={handleToggle}
            />
        </div>
    );
};
