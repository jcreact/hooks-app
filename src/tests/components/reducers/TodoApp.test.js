import React from 'react';
import { act } from '@testing-library/react';
import { shallow, mount } from 'enzyme';


// Componente
import { ToDoApp } from '../../../components/reducers/ToDoApp.jsx';
import { demoTodos } from '../../fixtures/demoTodos.js';

describe(`Pruebas en la aplicación 'ToDoApp.jsx'.`, () => {

    beforeEach(() => {
        Storage.prototype.setItem = jest.fn((k, v) => { });
        Storage.prototype.getItem = jest.fn((k) => null);
    });

    test(`Debe renderizar la aplicación 'TodoApp.jsx' correctamente.`, () => {
        const app = shallow(<ToDoApp />);
        expect(app).toMatchSnapshot();
        expect(app.find('ToDoAdd').exists()).toBeTruthy();
        expect(app.find('ToDoList').exists()).toBeTruthy();
        expect(app.find('span.badge').text()).toBe('0');
    });

    test('Debe agregar los todos de prueba y usar el LocalStorage.', () => {

        const app = mount(<ToDoApp />);

        let pending = 0;
        act(() => {
            demoTodos.forEach(t => {
                app.find('ToDoAdd').prop('handleAdd')(t);
                pending += (t.done ? 0 : 1);
            });
        });

        expect(app.find('span.badge').text()).toBe(pending.toString());
        expect(localStorage.setItem).toHaveBeenCalledTimes(2);
        expect(localStorage.setItem).toHaveBeenNthCalledWith(1, 'todos', '[]');
        expect(localStorage.setItem).toHaveBeenNthCalledWith(2, 'todos', JSON.stringify(demoTodos));
    });

    test(`Debe eliminar un todo.`, () => {
        const app = shallow(<ToDoApp />);

        app.find('ToDoAdd').prop('handleAdd')(demoTodos[0]);
        expect(app.find('span.badge').text()).toBe('1');

        app.find('ToDoList').prop('removeTask')(demoTodos[0].id);
        expect(app.find('span.badge').text()).toBe('0');
    });

});
