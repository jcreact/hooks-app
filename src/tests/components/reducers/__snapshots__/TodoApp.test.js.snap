// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en la aplicación 'ToDoApp.jsx'. Debe renderizar la aplicación 'TodoApp.jsx' correctamente. 1`] = `
<div
  className="container-fluid"
>
  <div
    className="mt-3 d-flex justify-content-start align-items-start"
  >
    <h1>
      To-Do App
    </h1>
    <span
      className="badge bg-danger rounded-pill"
    >
      0
    </span>
  </div>
  <hr />
  <ToDoAdd
    handleAdd={[Function]}
  />
  <ToDoList
    handleToggle={[Function]}
    removeTask={[Function]}
    todos={Array []}
  />
</div>
`;
